packages/flutter/lib/src/painting/decoration_image.dart

```diff
@@ -456,14 +456,64 @@ void paintImage({
     assert(sourceSize == inputSize, 'centerSlice was used with a BoxFit that does not guarantee that the image is fully visible.');
   }
 
-  // Output size is fully calculated.
+  if (repeat != ImageRepeat.noRepeat && destinationSize == outputSize) {
+    // There's no need to repeat the image because we're exactly filling the
+    // output rect with the image.
+    repeat = ImageRepeat.noRepeat;
+  }
+  final Paint paint = Paint()..isAntiAlias = isAntiAlias;
+  if (colorFilter != null)
+    paint.colorFilter = colorFilter;
+  if (sourceSize != destinationSize) {
+    paint.filterQuality = filterQuality;
+  }
+  paint.invertColors = invertColors;
+  final double halfWidthDelta = (outputSize.width - destinationSize.width) / 2.0;
+  final double halfHeightDelta = (outputSize.height - destinationSize.height) / 2.0;
+  final double dx = halfWidthDelta + (flipHorizontally ? -alignment.x : alignment.x) * halfWidthDelta;
+  final double dy = halfHeightDelta + alignment.y * halfHeightDelta;
+  final Offset destinationPosition = rect.topLeft.translate(dx, dy);
+  final Rect destinationRect = destinationPosition & destinationSize;
+
+  // Set to true if we added a saveLayer to the canvas to invert/flip the image.
+  bool invertedCanvas = false;
+  // Output size and destination rect are fully calculated.
   if (!kReleaseMode) {
     final ImageSizeInfo sizeInfo = ImageSizeInfo(
       // Some ImageProvider implementations may not have given this.
       source: debugImageLabel ?? '<Unknown Image(${image.width}×${image.height})>',
       imageSize: Size(image.width.toDouble(), image.height.toDouble()),
       displaySize: outputSize,
     );
+    assert(() {
+      if (debugInvertOversizedImages &&
+          sizeInfo.decodedSizeInBytes > sizeInfo.displaySizeInBytes + debugImageOverheadAllowance) {
+        final int overheadInKilobytes = (sizeInfo.decodedSizeInBytes - sizeInfo.displaySizeInBytes) ~/ 1024;
+        FlutterError.reportError(FlutterErrorDetails(
+          exception: 'Image $debugImageLabel has a display size of '
+            '${outputSize.width.toInt()}×${outputSize.height.toInt()} but a '
+            'decode size of ${image.width}×${image.height}, which '
+            'uses an additional ${overheadInKilobytes}kb.',
+          library: 'painting library',
+          context: ErrorDescription('while painting an image'),
+        ));
+        canvas.saveLayer(
```

Comment:
Consider adding some comments here about what each of these steps is doing.

Reply:
Done